pkg/stacks/unpack.go

```diff
@@ -17,320 +17,439 @@ limitations under the License.
-	_, err = os.Stdout.WriteString(content)
+	GotApp() bool
 
-	return err
+	AddGroup(string, StackGroup)
+	AddResource(string, StackResource)
+	AddIcon(string, v1alpha1.IconSpec)
+	AddUI(string, string)
+	AddCRD(string, *apiextensions.CustomResourceDefinition)
+
+	Yaml() string
 }
 
-func findRegistryRoot(fs afero.Fs, dir string) string {
-	if _, err := fs.Stat(filepath.Join(dir, registryDirName)); err == nil {
-		// the .registry subdir exists under the given dir, that must be the registry root
-		return filepath.Join(dir, registryDirName)
-	}
+// StackPackage defines the artifact structure of Stacks
+// A fully processed Stack can be thought of as a Stack CR (with its install and RBAC fields defined) and a collection of managed CRDs (annotated by their Stack resource, icon, group, and UI descriptors)
+type StackPackage struct {
+	// Stack is the Kubernetes API Stack representation
+	Stack v1alpha1.Stack
+
+	// CRDs are a map of the CRD files contained within a Stack indexed by their GVK
+	CRDs map[schema.GroupVersionKind]apiextensions.CustomResourceDefinition
+
+	// CRDPaths are used to map CRDs to their resource, icons, and group by directory name during post-processing
+	CRDPaths map[schema.GroupVersionKind]string
+
+	// Groups are Stack resource groups indexed by their filepath relative to the root of the Stack directory.
+	Groups map[string]StackGroup
+
+	// Icons are Stack resource icons indexed by their filepath relative to the root of the Stack directory
+	Icons map[string]*v1alpha1.IconSpec
+
+	// Resources are Stack resource metadata indexed by their filepath relative to the root of the Stack directory
+	Resources map[string]StackResource
+
+	// UISpecs are Stack UI Spec resources indexed by their filepath relative to the root of the Stack directory
+	UISpecs map[string]string
 
-	// didn't find a .registry subdir, the given dir must already be the root
-	return dir
+	// appSet indicates if a App has been assigned through SetApp for use by GotApp
+	appSet bool
 }
 
-func doUnpack(fs afero.Fs, root string) (string, error) { // nolint:gocyclo
-	// NOTE(displague): Adding Job put this over the cyclomatic complexity threshold.
-	// consider refactoring before adding more kinds
-	var output strings.Builder
+// Yaml represents the Stack Package as a multi-document YAML file containing the Stack and all CRDs contained within that Stack
+func (sp *StackPackage) Yaml() string {
+	var builder strings.Builder
 
-	// create a Stack record and populate it with the relevant package contents
-	v, k := v1alpha1.StackGroupVersionKind.ToAPIVersionAndKind()
-	stackRecord := &v1alpha1.Stack{TypeMeta: metav1.TypeMeta{APIVersion: v, Kind: k}}
+	builder.WriteString(yamlSeparator)
 
-	// find all CRDs and add to the stack record and the output builder
-	crdList, crdContent, err := readResources(fs, root)
-	if err != nil {
-		return "", err
+	for _, crd := range sp.CRDs {
+		b, err := yaml.Marshal(crd)
+		if err != nil {
+			log.Error(err, "Unpacking stack")
```

Comment:
this error is actually for marshaling a CRD while rendering the StackPackage to yaml, so probably update that.

feel free also to be more expressive with the error statements, for example saying which CRD this was that failed to marshal.  e.g.,
```go
for gvk, crd := range sp.CRDs {
		b, err := yaml.Marshal(crd)
		if err != nil {
			log.Error(err, "Marshaling CRD", "CRD GroupVersionKind", gvk)
```

Reply:
[Resolved](https://github.com/crossplaneio/crossplane/compare/9f5ab3c6d421d5e1b0d722af3317569614d5132f..017c3d47090daca29f82dfd30b95578a3b3e297d)